Realm/RLMQueryUtil.mm

```diff
@@ -113,6 +141,68 @@ void set_base_table(const Table*) override {}
     }
 };
 
+
+// Equal and ContainsSubstring are used by QueryBuilder::add_string_constraint as the comparator
+// for performing diacritic-insensitive comparisons.
+
+template <NSUInteger options>
+struct Equal {
+    using CaseSensitive = Equal<options & ~NSCaseInsensitiveSearch>;
+    using CaseInsensitive = Equal<options | NSCaseInsensitiveSearch>;
+
+    bool operator()(StringData v1, StringData v2, bool v1_null, bool v2_null) const
+    {
+        REALM_ASSERT_DEBUG(v1_null == v1.is_null());
+        REALM_ASSERT_DEBUG(v2_null == v2.is_null());
+
+        if (v1_null || v2_null) {
+            return v1_null == v2_null;
+        }
+
+        @autoreleasepool {
+            NSString *s1 = [[NSString alloc] initWithBytesNoCopy:(void*)v1.data() length:v1.size() encoding:NSUTF8StringEncoding freeWhenDone:NO];
+            NSString *s2 = [[NSString alloc] initWithBytesNoCopy:(void*)v2.data() length:v2.size() encoding:NSUTF8StringEncoding freeWhenDone:NO];
+
+            return [s1 compare:s2 options:options] == NSOrderedSame;
+        }
+    }
+};
+
+template <NSUInteger options>
+struct ContainsSubstring {
+    using CaseSensitive = ContainsSubstring<options & ~NSCaseInsensitiveSearch>;
+    using CaseInsensitive = ContainsSubstring<options | NSCaseInsensitiveSearch>;
+
+    bool operator()(StringData v1, StringData v2, bool v1_null, bool v2_null) const
+    {
+        REALM_ASSERT_DEBUG(v1_null == v1.is_null());
```

Comment:
Could avoid emitting a whole bunch of nearly identical copies of this function by moving the body to a function which takes `options` as a non-template parameter.